### PR TITLE
[WIP] Make pre-commit happy

### PR DESCRIPTION
**Pull request by bissenbay**
Fixes: https://github.com/thoth-station/python/issues/294

### CODE REVIEW

**Comment by fridex** [comment]
> ```python
>     from .base import PythonTestCase
> E   ImportError: attempted relative import with no known parent package
> ```

Try changing this to:

```python
from base import PythonTestCase
```

File: tests/test_aiosource.py [hunk 1]
```diff
@@ -18,32 +18,21 @@
...
-from thoth.python.aiosource import AIOSource, AsyncIterablePackages, AsyncIterableVersions
-from thoth.python.artifact import Artifact
 
-from .base import PythonTestCase
+from thoth.python.aiosource import AIOSource, AsyncIterablePackages
 
-import tempfile
-
-from zipfile import ZipFile
+from base import PythonTestCase
```
Comment (harshad16) [inline tests/test_aiosource.py:25]:
```suggestion
from tests.base import PythonTestCase
```
base is not a packages, as it a file from which we are importing, if we dont want to use `.base` then please use absolute path `tests.base`.
this is causing some of the tests to fail. please fix this in other test scripts.

File: thoth/python/__init__.py [hunk 2]
```diff
@@ -17,18 +17,6 @@
...
-from .digests_fetcher import DigestsFetcherBase
-from .digests_fetcher import PythonDigestsFetcher
-from .packages import Packages
-from .package_version import PackageVersion
-from .pipfile import Pipfile
-from .pipfile import PipfileMeta
-from .pipfile import PipfileLock
-from .project import Project
-from .source import Source
-from .aiosource import AIOSource, AsyncIterablePackages, AsyncIterableVersions, AsyncIterableArtifacts
```
Comment (harshad16) [inline thoth/python/__init__.py:29]: please dont remove them, add ether F401 or use `__all__` to let flake8 know about this.
if these imports are removed as a user of python package. 
i would have to use `thoth-python.digests_fetcher import DigestsFetcherBase` while i could use just `from thoth-python import DigestsFetcherBase`

some reference: https://stackoverflow.com/questions/31079047/python-pep8-class-in-init-imported-but-not-used

File: tests/test_aiosource.py (hunk 1 above)
Comment (bissenbay) [reply tests/test_aiosource.py]: did them

File: setup.py [hunk 3]
```diff
@@ -1,18 +1,22 @@
 #!/usr/bin/env python3
 
+"""Setup for python module."""
+
 import os
 import sys
 from setuptools import setup
 from pathlib import Path
-from setuptools.command.test import test as TestCommand
+from setuptools.command.test import test as TestCommand  # noqa: N812
```
Comment (harshad16) [inline setup.py:9]: Any specific reason for ignoring the N812 code here?

File: thoth/python/__init__.py (hunk 2 above)
Comment (bissenbay) [reply thoth/python/__init__.py:29]: thanks, they're back

File: setup.py (hunk 3 above)
Comment (bissenbay) [reply setup.py:9]: otherwise it pytest states `N812 lowercase 'test' imported as non lowercase 'TestCommand'`

Comment (harshad16) [reply setup.py:9]: so why not change the `TestCommand` to something like `test_command` and use it wherever necessary?

**Comment by bissenbay** [comment]
@fridex @harshad16 
Either `from base import PythonTestCase` and `from tests.base import PythonTestCase` seem to be not fixing `pytest` issues.

**Comment by harshad16** [comment]
> @fridex @harshad16
> Either `from base import PythonTestCase` and `from tests.base import PythonTestCase` seem to be not fixing `pytest` issues.

If you check the pytest error it doesn't  fail any more due to base import, there are some other import errors.

**Comment by bissenbay** [comment]
> > @fridex @harshad16
> > Either `from base import PythonTestCase` and `from tests.base import PythonTestCase` seem to be not fixing `pytest` issues.
> 
> If you check the pytest error it doesn't fail any more due to base import, there are some other import errors.

Should this imports `from .packages import Packages` be changed something like `from packages import Packages`?

**Comment by harshad16** [comment]
> > > @fridex @harshad16
> > > Either `from base import PythonTestCase` and `from tests.base import PythonTestCase` seem to be not fixing `pytest` issues.
> > 
> > 
> > If you check the pytest error it doesn't fail any more due to base import, there are some other import errors.
> 
> Should this imports `from .packages import Packages` be changed something like `from packages import Packages`?

i dont think , that would solve this issue, as packages is not a module package.
i will look into it.

File: thoth/python/package_version.py [hunk 4]
```diff
@@ -34,6 +33,10 @@
 from .exceptions import InternalError
 from .source import Source
 
+from thoth.python.pipfile import PipfileMeta
```
Comment (harshad16) [inline thoth/python/package_version.py:36]: This is causing the whole issue of PackageVersion cant be imported. 
also this is not used anywhere in here, please remove this.

**Comment by harshad16** [comment]
@bissenbay , based on the changes suggested , i think all the import error would be fixed in pytest.
however i notice the pytest would fail for lot of cases as `variable type list` is changed to something and the tests are not updated based on those changes.

i would suggest, just run the pytest in local and fix them
```
pipenv install --dev
pipenv run python setup.py test
```

File: thoth/python/package_version.py (hunk 4 above)
Comment (bissenbay) [reply thoth/python/package_version.py:36]: Harshad, I think it's being used, for example in this line: https://github.com/thoth-station/python/blob/master/thoth/python/package_version.py#L184, but there's a spelling error, it's written as `PipenvMeta` and it needs to be `PipfileMeta`.

File: setup.py (hunk 3 above)
Comment (bissenbay) [reply setup.py:9]: fixed

File: thoth/python/package_version.py (hunk 4 above)
Comment (bissenbay) [reply thoth/python/package_version.py:36]: If we remove that, then `pre-commit` won't be happy:
```
thoth/python/package_version.py:188: error: Name 'PipfileMeta' is not defined
thoth/python/package_version.py:208: error: Name 'PipfileMeta' is not defined
thoth/python/package_version.py:293: error: Name 'PipfileMeta' is not defined
Found 3 errors in 1 file (checked 12 source files)

black....................................................................Passed
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

thoth/python/package_version.py:188:36: F821 undefined name 'PipfileMeta'
thoth/python/package_version.py:208:91: F821 undefined name 'PipfileMeta'
thoth/python/package_version.py:293:98: F821 undefined name 'PipfileMeta'
```

**Comment by bissenbay** [comment]
@fridex @harshad16 
If we look carefully at let's say only one error, for example, `aiosource_test.py`:
```
============================================================================================== ERRORS ==============================================================================================
_____________________________________________________________________________ ERROR collecting tests/aiosource_test.py _____________________________________________________________________________
ImportError while importing test module '/home/bdauletb/Desktop/python/tests/aiosource_test.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../.local/share/virtualenvs/python-U1qOJFba/lib/python3.6/site-packages/_pytest/python.py:552: in _importtestmodule
    mod = import_path(self.fspath, mode=importmode)
../../.local/share/virtualenvs/python-U1qOJFba/lib/python3.6/site-packages/_pytest/pathlib.py:520: in import_path
    importlib.import_module(module_name)
/usr/lib64/python3.6/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
<frozen importlib._bootstrap>:994: in _gcd_import
    ???
<frozen importlib._bootstrap>:971: in _find_and_load
    ???
<frozen importlib._bootstrap>:955: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:665: in _load_unlocked
    ???
../../.local/share/virtualenvs/python-U1qOJFba/lib/python3.6/site-packages/_pytest/assertion/rewrite.py:170: in exec_module
    exec(co, module.__dict__)
tests/aiosource_test.py:23: in <module>
    from thoth.python.aiosource import AIOSource  # type: ignore
thoth/python/__init__.py:22: in <module>
    from .packages import Packages
thoth/python/packages.py:25: in <module>
    from .package_version import PackageVersion
thoth/python/package_version.py:36: in <module>
    from .pipfile import PipfileMeta
thoth/python/pipfile.py:35: in <module>
    from .packages import Packages
E   ImportError: cannot import name 'Packages'
```
I feel like, there's a loop happening in the import and that might be the reason why it's failing.
If we look carefully, we can see that it started with
```
thoth/python/__init__.py:22: in <module>
    from .packages import Packages
```
and came back to it
```
thoth/python/pipfile.py:35: in <module>
    from .packages import Packages
E   ImportError: cannot import name 'Packages'
```

File: thoth/python/package_version.py (hunk 4 above)
Comment (harshad16) [reply thoth/python/package_version.py]: ah, if it's being used then this going into looping import issue, we would have to rewrite and move around components.
Also, why is the PipfileMeta in quotes in line https://github.com/thoth-station/python/blob/master/thoth/python/package_version.py#L184 ?

Comment (bissenbay) [reply thoth/python/package_version.py]: Harshad, I have no idea, i think it works either way. First when we explicitly import the class and its name and maybe second way is the current way where we write it in quotes, but not sure.

**Comment by bissenbay** [comment]
@goern @fridex @harshad16 
I'm blocked, I have tried to solve this, but couldn't do anything more.

**Comment by fridex** [comment]
> @goern @fridex @harshad16
> I'm blocked, I have tried to solve this, but couldn't do anything more.

Has the circular import been resolved?

**Comment by bissenbay** [comment]
> > @goern @fridex @harshad16
> > I'm blocked, I have tried to solve this, but couldn't do anything more.
> 
> Has the circular import been resolved?

Yes, I think it was resolved and we are left with a bunch of `incompatible return type` errors.

**Comment by harshad16** [comment]
aicoe-ci is under maintenance, will be back in a few minutes.
thanks for your patience.

**Comment by harshad16** [comment]
The error logs is too large for ci to send in a comment.
It can be found out here: http://tekton-dashboard-tekton-pipelines.apps.thoth01.lab.upshift.rdu2.redhat.com/#/namespaces/thoth-ci/pipelineruns/aicoe-pipelinerun-p6grt
please check out the pytest step.

a snippet of it:
```
=========================== short test summary info ============================
FAILED tests/aiosource_test.py::TestAIOSource::test_get_packages - AssertionE...
FAILED tests/aiosource_test.py::TestAIOSource::test_get_package_versions_warehouse
FAILED tests/helpers_test.py::TestHelpers::test_parse_requirements - TypeErro...
FAILED tests/package_version_test.py::TestPackageVersion::test_from_pipfile_entry[tensorflow-package_info0-expected_values0-True]
FAILED tests/package_version_test.py::TestPackageVersion::test_from_pipfile_entry[tensorflow-package_info1-expected_values1-False]
FAILED tests/package_version_test.py::TestPackageVersion::test_from_pipfile_entry[tensorflow-package_info2-expected_values2-False]
FAILED tests/package_version_test.py::TestPackageVersion::test_parse_semver
FAILED tests/package_version_test.py::TestPackageVersion::test_parse_semver_leading_zeros
FAILED tests/package_version_test.py::TestPackageVersion::test_sorted - TypeE...
FAILED tests/package_version_test.py::TestPackageVersion::test_semver_error
FAILED tests/package_version_test.py::TestPackageVersion::test_normalize_python_package_name
FAILED tests/package_version_test.py::TestPackageVersion::test_normalize_python_package_version
FAILED tests/pipfile_test.py::TestPipfile::test_from_string[Pipfile_test1] - ...
FAILED tests/pipfile_test.py::TestPipfile::test_pipfile_extras_parsing - Type...
FAILED tests/pipfile_test.py::TestPipfile::test_construct_requirements - Type...
FAILED tests/pipfile_test.py::TestPipfileLock::test_from_string[Pipfile_test1.lock]
FAILED tests/pipfile_test.py::TestPipfileLock::test_extras_parsing - TypeErro...
FAILED tests/project_test.py::TestProject::test_add_package - TypeError: Type...
FAILED tests/project_test.py::TestProject::test_add_package_develop - TypeErr...
FAILED tests/project_test.py::TestProject::test_add_source - TypeError: Type ...
FAILED tests/project_test.py::TestProject::test_get_outdated_package_versions_indirect
FAILED tests/project_test.py::TestProject::test_get_outdated_package_versions_direct
FAILED tests/project_test.py::TestProject::test_indexes_in_meta - TypeError: ...
FAILED tests/project_test.py::TestProject::test_from_pip_compile_files_example_dir2
FAILED tests/project_test.py::TestProject::test_from_pip_compile_files_example_dir1
FAILED tests/project_test.py::TestProject::test_from_pip_compile_files_example_dir3
FAILED tests/provenance_test.py::TestProvenanceCheck::test_different_artifacts_on_sources[index_report0]
FAILED tests/provenance_test.py::TestProvenanceCheck::test_different_source_error[index_report0]
FAILED tests/provenance_test.py::TestProvenanceCheck::test_possible_different_source_info[index_report0]
FAILED tests/provenance_test.py::TestProvenanceCheck::test_missing_package_error[index_report0]
FAILED tests/provenance_test.py::TestProvenanceCheck::test_invalid_artifact_hash_error[index_report0
```

**Comment by goern** [comment]
is this still WIP?

**Comment by bissenbay** [comment]
> is this still WIP?

Yes

**Comment by bissenbay** [comment]
https://github.com/thoth-station/python/pull/314

**Pull request closed without merging**